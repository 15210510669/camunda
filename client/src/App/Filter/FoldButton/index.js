export {default as FoldButton} from './FoldButton';
