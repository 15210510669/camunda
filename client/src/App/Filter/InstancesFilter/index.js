export {default as InstancesFilter} from './InstancesFilter';
