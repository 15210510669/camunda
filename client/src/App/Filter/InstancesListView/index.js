export {default as InstancesListView} from './InstancesListView';
