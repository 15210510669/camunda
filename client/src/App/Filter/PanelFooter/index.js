export {default as PanelFooter} from './PanelFooter';
