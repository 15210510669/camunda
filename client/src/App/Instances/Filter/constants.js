export const LABELS = {
  RUNNING: 'Running Instances',
  ACTIVE: 'Active',
  INCIDENTS: 'Incidents',
  FINISHED: 'Completed Instances',
  COMPLETED: 'Regularly Completed',
  CANCELED: 'Canceled'
};
