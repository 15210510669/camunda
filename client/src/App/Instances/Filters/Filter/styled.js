import styled from 'styled-components';

export const Filters = styled.div`
  padding: 10px 0;
  display: flex;
  flex-direction: column;
`;

export const NestedFilters = styled.div`
  padding-left: 26px;
`;
