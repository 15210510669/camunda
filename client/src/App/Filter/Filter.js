import React, {Component} from 'react';
import update from 'immutability-helper';

import Header from '../Header';

import Panel from 'modules/components/Panel';

import InstancesFilter from './InstancesFilter';
import InstancesListView from './InstancesListView';

import {getCount} from './api';

import * as Styled from './styled.js';

export default class Filter extends Component {
  state = {
    filter: {running: true},
    filterCount: 0,
    selection: {
      list: new Set(),
      isBlacklist: false
    }
  };

  handleFilterChange = async change => {
    const filter = update(this.state.filter, change);
    this.setState({
      filter
    });
    // separate setState to not block UI while waiting for server response
    this.setState({
      filterCount: await getCount(filter)
    });
  };

  getSelectionCount = () => {
    const {
      filterCount,
      selection: {list, isBlacklist}
    } = this.state;

    return isBlacklist ? filterCount - list.size : list.size;
  };

  render() {
    return (
      <div>
        <Header
          active="instances"
          instances={14576}
          filters={this.state.filterCount}
          selections={this.getSelectionCount()}
          incidents={328}
        />
        <Styled.Filter>
          <Styled.Left>
            <Panel isRounded>
              <Panel.Header isRounded foldButtonType="left">
                Filters
              </Panel.Header>
              <Panel.Body>
                <InstancesFilter
                  filter={this.state.filter}
                  onChange={this.handleFilterChange}
                />
              </Panel.Body>
              <Panel.Footer />
            </Panel>
          </Styled.Left>
          <Styled.Right>
            <Styled.Top>
              <Panel isRounded>
                <Panel.Header isRounded>Process Definition Name</Panel.Header>
                <Panel.Body>Process Definition Name content</Panel.Body>
              </Panel>
            </Styled.Top>
            <Styled.Bottom>
              <InstancesListView
                instancesInFilter={this.state.filterCount}
                onSelectionUpdate={change => {
                  this.setState({
                    selection: update(this.state.selection, change)
                  });
                }}
                selection={this.state.selection}
                filter={this.state.filter}
              />
            </Styled.Bottom>
          </Styled.Right>
        </Styled.Filter>
      </div>
    );
  }

  async componentDidMount() {
    this.setState({
      filterCount: await getCount(this.state.filter)
    });
  }
}
