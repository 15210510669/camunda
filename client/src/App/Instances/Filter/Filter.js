import React from 'react';
import PropTypes from 'prop-types';

import Checkbox from 'modules/components/Checkbox';
import {LABELS} from 'modules/constants/filter';

import * as Styled from './styled.js';

export default class Filter extends React.Component {
  static propTypes = {
    filter: PropTypes.shape({
      active: PropTypes.bool,
      incidents: PropTypes.bool,
      canceled: PropTypes.bool,
      completed: PropTypes.bool
    }).isRequired,
    onChange: PropTypes.func.isRequired,
    type: PropTypes.oneOf(['running', 'finished']).isRequired
  };
  constructor(props) {
    super(props);
    this.childFilterTypes = Object.keys(props.filter);
  }

  getCheckedChildrenCount = () => {
    return Object.values(this.props.filter).filter(value => value).length;
  };

  isIndeterminate = () => {
    return this.getCheckedChildrenCount() === 1;
  };

  handleChange = type => async () => {
    const {filter} = this.props;
    const newFilter = {
      [type]: filter[type] ? !filter[type] : true
    };

    this.props.onChange(newFilter);
  };

  onResetFilter = () => {
    const change = {};
    const bothChecked = this.getCheckedChildrenCount() === 2;

    this.childFilterTypes.map(type =>
      Object.assign(change, {
        [type]: bothChecked ? !this.props.filter[type] : true
      })
    );

    this.props.onChange(change);
  };

  render() {
    const {type, filter} = this.props;
    return (
      <Styled.Filters>
        <div>
          <Checkbox
            label={LABELS[type]}
            isIndeterminate={this.isIndeterminate()}
            isChecked={this.getCheckedChildrenCount() === 2}
            onChange={this.onResetFilter}
          />
        </div>
        <Styled.NestedFilters>
          {this.childFilterTypes.map(type => {
            return (
              <Checkbox
                key={type}
                label={LABELS[type]}
                isChecked={filter[type] || false}
                onChange={this.handleChange(type)}
              />
            );
          })}
        </Styled.NestedFilters>
      </Styled.Filters>
    );
  }
}
