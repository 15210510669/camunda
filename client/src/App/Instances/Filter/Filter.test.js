import React from 'react';
import {shallow} from 'enzyme';

import Filter from './';

describe('Filter', () => {
  let node;
  const mockOnChange = jest.fn();
  let active = false;
  let incidents = false;
  beforeEach(() => {
    node = shallow(
      <Filter
        type={'running'}
        filter={{active, incidents}}
        onChange={mockOnChange}
      />
    );
  });

  it('should render three checkboxes', () => {
    expect(node.children().find('Checkbox').length).toBe(3);
  });

  it('should render the Checkboxes "checked" according to the filter props', () => {
    //given
    const activityFilter = false;
    const incidentsFilter = false;
    expect(node.instance().props.filter.active).toBe(activityFilter);
    expect(node.instance().props.filter.incidents).toBe(incidentsFilter);
    //then
    expect(
      node
        .childAt(1)
        .childAt(1)
        .find('Checkbox')
        .props().isChecked
    ).toBe(incidentsFilter);

    expect(
      node
        .childAt(1)
        .childAt(0)
        .find('Checkbox')
        .props().isChecked
    ).toBe(activityFilter);
  });

  it('parent checkbox should be indetermniate', () => {
    let node;
    const mockOnChange = jest.fn();
    let active = false;
    let incidents = true;

    node = shallow(
      <Filter
        type={'running'}
        filter={{active, incidents}}
        onChange={mockOnChange}
      />
    );
    expect(node.childAt(0).find('Checkbox')).toMatchSnapshot();
  });

  it('parent checkbox should be unchecked', () => {
    let node;
    const mockOnChange = jest.fn();
    let active = false;
    let incidents = false;

    node = shallow(
      <Filter
        type={'running'}
        filter={{active, incidents}}
        onChange={mockOnChange}
      />
    );
    expect(node.childAt(0).find('Checkbox')).toMatchSnapshot();
  });
});
