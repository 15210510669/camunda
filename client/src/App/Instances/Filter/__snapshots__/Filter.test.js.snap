// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter parent checkbox should be indetermniate 1`] = `
<Checkbox
  isChecked={false}
  isIndeterminate={true}
  label="Running Instances"
  onChange={[Function]}
/>
`;

exports[`Filter parent checkbox should be unchecked 1`] = `
<Checkbox
  isChecked={false}
  isIndeterminate={false}
  label="Running Instances"
  onChange={[Function]}
/>
`;
